src/Elastic.Apm/DiagnosticSource/DiagnosticInitializer.cs

```diff
@@ -18,7 +18,10 @@ public void OnNext(DiagnosticListener value)
 		{
 			foreach (var listener in _listeners)
 			{
-				if (value.Name == listener.Name) value.Subscribe(listener);
+				if (value.Name != listener.Name) continue;
+
+				var sourceSubscription = value.Subscribe(listener);
+				listener.SourceSubscription = sourceSubscription;
```

Comment:
I think it would be better if `DiagnosticInitializer` implements `IDisposable` and manages its own state that way. 

`DiagnosticsInitializer` behaves a bit like `Subject<T>` in that its both an observer of `DiagnosticListener` while subscribing to `IObservable<KeyValuePair<string, string>>`. It might be worth exploring making that more explicit.

This would put `DiagnosticsInitializer` in the driving seat without "tagging" the observers with a disposable.

Reply:
Thx, I adapted this accordingly.